google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/gaxx/grpc/BigtableChannelPool.java

```diff
@@ -138,15 +140,94 @@ public String authority() {
   }
 
   /**
-   * Create a {@link ClientCall} on a Channel from the pool chosen in a round-robin fashion to the
-   * remote operation specified by the given {@link MethodDescriptor}. The returned {@link
-   * ClientCall} does not trigger any remote behavior until {@link
-   * ClientCall#start(ClientCall.Listener, io.grpc.Metadata)} is invoked.
+   * Create a {@link ClientCall} on a Channel from the pool to the remote operation specified by the
+   * given {@link MethodDescriptor}. The returned {@link ClientCall} does not trigger any remote
+   * behavior until {@link ClientCall#start(ClientCall.Listener, io.grpc.Metadata)} is invoked.
    */
   @Override
   public <ReqT, RespT> ClientCall<ReqT, RespT> newCall(
       MethodDescriptor<ReqT, RespT> methodDescriptor, CallOptions callOptions) {
-    return getChannel(indexTicker.getAndIncrement()).newCall(methodDescriptor, callOptions);
+    return new AffinityChannel(pickEntryIndex()).newCall(methodDescriptor, callOptions);
+  }
+
+  /** Pick an entry to use for the next call. */
+  private int pickEntryIndex() {
+    switch (settings.getLoadBalancingStrategy()) {
+      case ROUND_ROBIN:
+        return pickEntryIndexRoundRobin();
+      case LEAST_IN_FLIGHT:
+        return pickEntryIndexLeastInFlight();
+      case POWER_OF_TWO_LEAST_IN_FLIGHT:
+        return pickEntryIndexPowerOfTwoLeastInFlight();
+      default:
+          LOG.warning(String.format(
+            "Unknown load balancing strategy %s, falling back to ROUND_ROBIN.",
+            settings.getLoadBalancingStrategy()));
+          return pickEntryIndexRoundRobin();
+
+    }
+  }
+
+  /** Pick an entry using the Round Robin algorithm. */
+  private int pickEntryIndexRoundRobin() {
+    return indexTicker.getAndIncrement();
+  }
+
+  /** Pick an entry at random. */
+  private int pickEntryIndexRandom() {
+    List<Entry> localEntries = entries.get();
+    for (int attempt = 0; attempt < 5; attempt++) {
+      int choice = rng.nextInt(localEntries.size());
+      if (localEntries.get(choice).retainable()) {
```

Comment:
I dont think prevents any race conditions - the entry can be shutdown between the call to retainable() and the channel being used. I believe that this should already be dealt with in getRetainedEntry() during newCall time.

Reply:
Also I think you want to re-fetch the entry list on each loop